File: adt2unit/pas/slide_modulator_volume_down.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

procedure slide_modulator_volume_down(chan: Byte; slide: Byte);

var
  vol: Word;
  vLo,vHi: Byte;

begin
  vLo := LO(volume_table[chan]);
  vHi := HI(volume_table[chan]);
  If (vLo+slide <= 63) then vol := concw(vLo+slide,vHi)
  else vol := concw(63,vHi);
  set_ins_volume(LO(vol),BYTE_NULL,chan);
  volume_table[chan] := vol;
end;

File: adt2unit/slide_modulator_volume_down.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

// chan: 0..19
static void slide_modulator_volume_down (uint8_t chan, uint8_t slide)
{
  uint8_t vLo = volume_table[chan] & 0xFF;
  uint8_t vHi = volume_table[chan] >> 8;

  vLo = ((vLo + slide) <= 63) ? vLo + slide : 63;

  set_ins_volume (vLo, BYTE_NULL, chan + 1);
  volume_table[chan] = vLo + (vHi << 8);
}
